db/db_versioned.go

```diff
@@ -7,47 +7,42 @@
 package db
 
 import (
+	"bytes"
 	"context"
+	"fmt"
+	"math"
 
+	"github.com/pkg/errors"
+	bolt "go.etcd.io/bbolt"
+
+	"github.com/iotexproject/iotex-core/v2/db/batch"
 	"github.com/iotexproject/iotex-core/v2/pkg/lifecycle"
+	"github.com/iotexproject/iotex-core/v2/pkg/util/byteutil"
+)
+
+var (
+	ErrDeleted = errors.New("deleted in DB")
+	_minKey    = []byte{0} // the minimum key, used to store namespace's metadata
 )
 
 type (
-	// KvVersioned is a versioned key-value store, where each key has multiple
-	// versions of value (corresponding to different heights in a blockchain)
-	//
-	// Versioning is achieved by using (key + 8-byte version) as the actual
-	// storage key in the underlying DB. For each bucket, a metadata is stored
-	// at the special key = []byte{0}. The metadata specifies the bucket's name
-	// and the key length.
-	//
-	// For each versioned key, the special location = key + []byte{0} stores the
-	// key's metadata, which includes the following info:
-	// 1. the version when the key is first created
-	// 2. the version when the key is lastly written
-	// 3. the version when the key is deleted
-	// 4. hash of the key's last written value (to detect/avoid same write)
-	// If the location does not store a value, the key has never been written.
-	//
-	// How to use a versioned DB:
-	//
-	// db := NewBoltDBVersioned(cfg) // creates a versioned DB
-	// db.Start(ctx)
-	// defer func() { db.Stop(ctx) }()
-	//
-	// kv := db.SetVersion(5)
-	// value, err := kv.Get("ns", key) // read 'key' at version 5
-	// kv = db.SetVersion(8)
-	// err := kv.Put("ns", key, value) // write 'key' at version 8
-
-	KvVersioned interface {
+	VersionedDB interface {
 		lifecycle.StartStopper
 
+		// Put insert or update a record identified by (namespace, key)
```

Comment:
Why to modify the interface this way? I would recommend that the version should be for all keys rather than individual keys. Do we have a use case that requires this? I think  the original interface design is easier for integration, or will there be a wrapper to encapsulate it?

Reply:
yes, `KvVersioned` is the new wrapper to encapsulate it